### PR TITLE
content/concepts/peer-routing.md: add first draft

### PR DESCRIPTION
**Pull request by poolsar42**
Hey

It's my first contribution to libp2p docs and libp2p in general - so please be kind 👀 

I'm writing this doc alongside my learning process - so if I am missing something, or writing something completely wrong - please let me know!

And before next draft I want to ask these questions:
- How long will it take to wait a ping to evict node? And should I mention this in the docs?
- How often bucket refresh is performing?
- Should I mention libp2p interface here? Or is it more appropriate in content-routing?
- (again) Let me know if I'm missing something!

Feel free to change everything here if you feel so. My main wish is to have complete docs 😋 

Thanks

### CODE REVIEW

**Comment by poolsar42** [comment]
@mxinden I've seen that you would like to help with the docs, so I'm pinging you :)

**Comment by poolsar42** [comment]
#13

**Comment by BigLep** [comment]
@mxinden : are you ok to take this?

File: content/concepts/peer-routing.md
Comment (mxinden) [inline content/concepts/peer-routing.md:30]: Mhh, these don't seem to render in GitHubs markdown view. Do they do so on your end?

File: content/concepts/peer-routing.md
Comment (mxinden) [inline content/concepts/peer-routing.md:6]: It is one mechanism among many. E.g. ethereum 2 is not using it at all.

File: content/concepts/peer-routing.md
Comment (mxinden) [inline content/concepts/peer-routing.md:16]:
```suggestion
In Kademlia peers are key-value pairs in the DHT. Conceptually peer ID stored as a key and [address](/concepts/addressing/) (or IP address) of a peer stored as a value.
```

File: content/concepts/peer-routing.md
Comment (mxinden) [inline content/concepts/peer-routing.md:18]: We actually hash the IDs with sha256 to get a uniform key space. That said, I am not sure it needs mentioning here.

File: content/concepts/peer-routing.md
Comment (mxinden) [inline content/concepts/peer-routing.md:58]: Cool to only introduce this complexity later on in the article :+1:

File: content/concepts/peer-routing.md
Comment (mxinden) [inline content/concepts/peer-routing.md:68]: Not quite. It will be marked as non-responding and eventually replaced once we have a better node.

File: content/concepts/peer-routing.md
Comment (mxinden) [inline content/concepts/peer-routing.md:76]: First thing I would link to is the Kademlia paper. Second would be the libp2p Kademlia specification.

- https://github.com/libp2p/specs/blob/master/kad-dht/README.md
- https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf

**Review by mxinden — COMMENTED** [review]
Thanks for the help @poolsar42.

@walkerlj0 @johnnymatthews as native speakers, does one of you have some time to give this a review from a language perspective?

**Comment by johnnymatthews** [comment]
> @walkerlj0 @johnnymatthews as native speakers, does one of you have some time to give this a review from a language perspective? - @mxinden 

For sure! Although, I'm gonna wait until the above conversations have been resolved. Ping me when stuff's good for a review.

**Comment by walkerlj0** [comment]
@mxinden @johnnymatthews @yusefnapora  I am working on taking a look at this, right now I cannnot build this version of Hugo on my M1, Yusef is working on helping update

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md:30]: No, they don't for me
I was doing the same thing as in [pubsub](https://github.com/libp2p/docs/blob/master/content/concepts/publish-subscribe.md) article, so it shows up [here](https://bafybeidklhiwzma7r3vufxkvoym3pq6pv2l2hnmnu5uynvpqfjnbqmd2bi.on.fleek.co/concepts/peer-routing/)

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md]: fixed in new commit, but still not sure if it sounds good, since I don't know yet many other ways to discover peers 🙃

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md]: 👍

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md]: I think, that it is good to mention this and then show hash is being stored as binary

That is because I also think that for the content-routing it would be good to explain distances

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md:58]: Oh I also like this, it wasn't originally my idea though

I borrowed it from [here](https://codethechange.stanford.edu/guides/guide_kademlia.html)
(and I referred to this source at the end :))

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md]: fixed

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md:76]: And added that

Comment (poolsar42) [reply content/concepts/peer-routing.md:76]: I should have guessed to add the specs :))

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:10]:
```suggestion
In this example, we are going to assume for now that our DHT is static, in hat one joins DHT and no one leaves.
```

**Comment by mxinden** [comment]
Good to go from a concept perspective. Input from @walkerlj0 and @johnnymatthews would be great.

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:24]:
```suggestion
The red node in this diagram has the prefix `001`. Traversing this graph from the top to the bottom along the edges, you will notice a blue node with the prefix `11101`, as well as yellow leaves which are other peers. 

For any given node, Kademlia divides the whole binary tree into a series of successively lower subtrees. The highest subtree consists of the half of the binary tree not containing the original node. The next subtree consists of the half of the remaining tree not containing the next highest node, and so on. 

You can see here for node `001`  there are three levels of subtrees are also circle,  consisting of all nodes with prefixes `1` (highest level subtree) , `01`, `000` (lowest level subtree).
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:30]: @poolsar42 I think you have to upload the asset into the correct folder, like these other images: https://github.com/libp2p/docs/tree/master/content/concepts/publish-subscribe

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:48]:
```suggestion
The place where one node stores info about other nodes is called a routing table. The routing table is a binary tree whose leaves are `k-buckets`. 
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:28]:
```suggestion
Let's see how once can locate in an example. We will call the red node 'Bob', with the prefix  `001.` We will call the yellow node Alice, with the prefix `01`

Bob can locate Alice in this example.  Bob knows Alice's peerID, but he needs to get Alice's location address. From the subtree with the prefix `1` Bob knows a peer with prefix `101`. He sends a message to their address, where he asks for Alice's address.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:32]:
```suggestion
Peer `100` doesn't have Alice's address, so they reroute and ask the peer located at the node with the prefix `110` from their subtree.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:36]:
```suggestion
In this example, peer `110` also doesn't have Alice's address, so they send request to another node, with the prefix `1111`.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [reply content/concepts/peer-routing.md:28]: I'm not entirely sure what you mean by this. The words address, key, location address keep being used interchangeably. I recommend you review all of this paragraph to accurately describe what is being located. Is it a PeerID? IP Address?

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:50]:
```suggestion
### k-buckets
Take a look at the initial tree:
 ![Kademlia-system](Kademlia-system.png) 
See this circled sub-trees? Remember that the Kademlia protocol ensures that every node knows of (and can locate) at least one node in each of its subtrees. If that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:60]:
```suggestion
### Peers Leaving 
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:72]:
```suggestion
### Peers Joining
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:66]:
```suggestion
And, when hopping, instead of asking one computer from the `k-bucket` we will be asking `k` number of peers. Some may not reply, but not all.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:62]:
```suggestion
Let us assume that we have some known parameter `k` that represents some given number of computers.  This number should be a value such that it is extremely unlikely that all of those computers will leave the network in the same hour. The original Kademlia paper that the `k` value should be 20.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:68]:
```suggestion
To check if each node is still alive we can ping it periodically. If node doesn't respond in the allotted time it will be marked as non-responding and eventually replace it with another.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:70]:
```suggestion
How do we become aware of new computers that could fit into our unfilled `k-buckets`? We can perform lookups on random ids within a `k-bucket’s` id range (the range is defined by all leaves that are direct descendants of the `k-bucket’s` corresponding internal node, which is not a leaf, in the complete tree) and thus learn about other computers within that `k-bucket`, if they exist. This is called a _bucket refresh_.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:74]:
```suggestion
To join the DHT new computer, which we will call `Eve`,  must know at least on IP address of a computer which is in the network. Then `Eve` will perform a lookup of herself to get closest known computers which are already aware of `Eve`. After that `Eve` will perform a _bucket refresh_ for all `k-buckets` from the closest known computers to `Eve`. 

This will populate the `Eve's` routing table based on the correctness of the lookup algorithm and the fact that the routing tables of other computers are sufficiently populated. In addition, all computers that are being queried by `Eve` will become aware of `Eve` in the process and thus will have the opportunity to insert `Eve` into their own `k-buckets` if their corresponding bucket is not already full.
```

**Review by walkerlj0 — CHANGES_REQUESTED** [review]
Hey, made several changes, most of which are adding articles (a, the), some minor language modifications, as well as a few re-writes & tense changes)

One thing I was confused on was the paragraphs about 'Alice' and her peer ID. You mention the words key, IP address, and location-- I'm not sure which is which, and if they refer to the same thing, perhaps just use one of those words each time (question- does Kademlia look up & provide IP addresses, keys, PeerIDs or all?)

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md:30]: @walkerlj0 I think I did the same thing: you can see that for publish-subscribe concept "publish-subscribe" was created, and I just created "peer-routing" folder for content for this article.
Please hmu if you feel that I'm doing it wrong, because [here](https://bafybeidklhiwzma7r3vufxkvoym3pq6pv2l2hnmnu5uynvpqfjnbqmd2bi.on.fleek.co/concepts/peer-routing/) it's rendering just fine (this is check for fleek build)

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md:28]: Yeah, i'm sorry for this mess.

Here DHT is a storage for key-value pairs. 
In our case key is Alice's peer ID. Bob knows it, and he needs to get the value.
Value here is Alice's IP address, so then Bob will be able to communicate directly with this IP address (and this is Alice)

I added "a brief reminder" paragraph in my current commit and I hope it'll make it clearer.

**Comment by poolsar42** [comment]
Thank you @walkerlj0! 

Hope I didn't miss anything and included all your suggestions.

Also I slightly changed paragraph with Alice's iP address, key and peer ID. Again, I hope it's clearer now. But I'm happy to change it more, if there are still unclear places there :)

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:28]:
```suggestion
You can see here for node `001` there are three levels of subtrees are also circle, consisting of all nodes with prefixes `1` (highest level subtree) , `01`, and `000` (lowest level subtree).
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:62]:
```suggestion
See these circled sub-trees? Remember that the Kademlia protocol ensures that every node knows of (and can locate) at least one node in each of its subtrees. If that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [inline content/concepts/peer-routing.md:70]:
```suggestion
The gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
```

File: content/concepts/peer-routing.md
Comment (walkerlj0) [reply content/concepts/peer-routing.md:38]: I was able to fetch this branch and render the images, (they look great and add so much to the article!) @poolsar42

File: content/concepts/peer-routing.md
Comment (poolsar42) [reply content/concepts/peer-routing.md]: Sorry for taking so much time 😅
And thanks!

Comment (walkerlj0) [reply content/concepts/peer-routing.md]: No worries at all, thanks for contributing and taking the time to write this!

**Comment by walkerlj0** [comment]
Hey @poolsar42, you can either 'commit suggestion' for all the edits I made, or else if you don't want that change, just click 'resolve conversation'.

Once you have done that for all comments, I can take another pass. See the video here of what I am talking about: https://drive.google.com/file/d/1ypc9wKYfN6AtgcEdynosZ24E3FeiFM_Q/view?usp=sharing

**Comment by poolsar42** [comment]
@walkerlj0 neat, thanks!
I've committed them all by hands 😅

File: content/concepts/peer-routing.md
Comment (MarcoPolo) [inline content/concepts/peer-routing.md:78]: I believe this isn't all `k` peers but some other number `α` of conucrrent reqs we make

**Comment by MarcoPolo** [comment]
First off, thank you @poolsar42 for taking the time to write this up and learning these concepts. I enjoyed reading it! 😄 

That said, I'm of the opinion that we shouldn't merge this into our docs. I think our docs for peer-routing shouldn't be as ambitious as this to try to explain Kademlia from the ground up. Instead it should give a very high level overview and link to other references where folks can learn more. Like linking to the original paper, implementations, or something like this write-up.

Instead I encourage you to take this content and publish it on your own blog (and maybe even host it on IPFS) and we'll gladly link to it (or a content addresses ipfs hash).

I'll give @mxinden and @marten-seemann some time to chime in with their thoughts here too.

Again, thanks for this and I hope you got some good feedback :)

File: content/concepts/peer-routing.md
Comment (MarcoPolo) [inline content/concepts/peer-routing.md:78]: I believe this isn't all k peers but some other number α of conucrrent reqs we make

**Comment by poolsar42** [comment]
@MarcoPolo thanks for feedback, it was so nice writing and correcting this all time :)

I'll of course publish it to my blog, if you don't accept PR here, and I'll send you the link to it.
I already have established blog hosted on IPFS with IPNS link, so it's not a CID (and I'm proud of it kinda), Fleek did all work for me :)

Thanks for ya'll for your comments and feedback, it was a pleasure

And I'm really sorry for not understanding issue #13 properly :(

**Pull request closed without merging**
